pkg/applicationserver/io/pubsub/provider/mqtt/driver.go

```diff
@@ -0,0 +1,280 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mqtt
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"time"
+
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/driver"
+)
+
+type topic struct {
+	client  mqtt.Client
+	topic   string
+	timeout time.Duration
+	qos     byte
+}
+
+var errNilClient = errors.DefineInvalidArgument("nil_client", "client is nil")
+
+// OpenTopic returns a *pubsub.Topic that publishes to the given topic name with the given MQTT client.
+func OpenTopic(client mqtt.Client, topicName string, timeout time.Duration, qos byte) (*pubsub.Topic, error) {
+	dt, err := openDriverTopic(client, topicName, timeout, qos)
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewTopic(dt, nil), nil
+}
+
+func openDriverTopic(client mqtt.Client, topicName string, timeout time.Duration, qos byte) (driver.Topic, error) {
+	if client == nil {
+		return nil, errNilClient
+	}
+	dt := &topic{
+		client:  client,
+		topic:   topicName,
+		timeout: timeout,
+		qos:     qos,
+	}
+	return dt, nil
+}
+
+// SendBatch implements driver.Topic.
+func (t *topic) SendBatch(ctx context.Context, msgs []*driver.Message) error {
+	if t == nil || t.client == nil {
+		return errNilClient
+	}
+	for _, msg := range msgs {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if msg.BeforeSend != nil {
+			asFunc := func(i interface{}) bool { return false }
+			if err := msg.BeforeSend(asFunc); err != nil {
+				return err
+			}
+		}
+		body, err := encodeMessage(msg)
+		if err != nil {
+			return err
+		}
+		if token := t.client.Publish(t.topic, t.qos, false, body); !token.WaitTimeout(t.timeout) {
+			return convertToCancelled(token.Error())
+		}
+	}
+	return nil
+}
+
+func encodeMessage(dm *driver.Message) ([]byte, error) {
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
+	if len(dm.Metadata) == 0 {
+		return dm.Body, nil
+	}
+	err := enc.Encode(dm.Metadata)
+	if err != nil {
+		return nil, err
+	}
+	err = enc.Encode(dm.Body)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func decodeMessage(message mqtt.Message) (*driver.Message, error) {
+	asFunc := func(i interface{}) bool {
+		p, ok := i.(*mqtt.Message)
+		if !ok {
+			return false
+		}
+		*p = message
+		return true
+	}
+	buf := bytes.NewBuffer(message.Payload())
+	dec := gob.NewDecoder(buf)
+	dm := &driver.Message{
+		AckID:  -1,
+		AsFunc: asFunc,
+	}
+	err := dec.Decode(&dm.Metadata)
+	if err != nil {
+		dm.Metadata = nil
+		dm.Body = message.Payload()
+		return dm, nil
+	}
+	return dm, dec.Decode(&dm.Body)
```

Comment:
I believe this works but is there a reference somewhere?

Reply:
This follows the same concept as the [NATS implementation](https://github.com/google/go-cloud/blob/master/pubsub/natspubsub/nats.go#L403-L428).

I kept the `encoding/gob` solution since using JSON encoding is not really possible (how do you make a difference between having a JSON body that is an envelope or not, when the envelope is JSON itself ?).

Reply:
I don't like this solution very much either, but since we not using metadata anyway it will not affect us.

Reply:
In which case do we have an envelope?

Reply:
In our pubsub provider we never use this envelope, but the [test harness](https://github.com/google/go-cloud/blob/master/pubsub/drivertest/drivertest.go) expects some kind of metadata support.